src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -947,4 +983,22 @@ public String concreteIndex() {
             return concreteIndex;
         }
     }
+
+    static class IndexShardReference {
```

Comment:
here is a suggestion, why don't we make this class implement Closeable and reduce the entire thing into a single method we can override ie:

```
protected Closeable newShardReference(ShardId shardId) {
  // get the index shard
  return new IndexShardReference(indexShard);
}

// example usage
public void messageReceived(final ReplicaOperationRequest request, final TransportChannel channel) throws Exception {
             try(Closeable indexShardReference = newShardReference(request.shardId)) {
                 shardOperationOnReplica(request);
             } catch (Throwable t) {
                 failReplicaIfNeeded(request.shardId.getIndex(), request.shardId.id(), t);
                 throw t;
             }
             channel.sendResponse(TransportResponse.Empty.INSTANCE);
         }
```

that way we can keep you incrementOperationCounter methods on IndexShard but don't need to mock all the stuff?